### DISCUSSION TITLE
Modifications to computeValue() in ArrayRankTwoAux.C

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#20957] [Q&A General] [answered]
I am working towards implementing the array auxvariables and auxkernels as suggested in [20213](https://github.com/idaholab/moose/discussions/20213). Looks like the function computeValue() #has to be maintained in a standard as-is manner, as any modifications to this function signature is resulting in a compiler error. Earlier i was trying to force this function to return a 2D vector (of size [48,6] where i store the elemental stresses and other identifiers needed for transfer class) instead of a scalar `Real` and that backfired me. I wanted to ask here how important is this function when we migrate to array auxkernels and whether it is a good idea to code that up in an another function leaving the computeValue to return just a `namesake` scalar value.

**Comment by GiudGiud** [comment, marked as answer]
Hello

So what is ArrayRankTwoAux inheriting from?
You want to make sure to inherit from an array auxkernel, as only those have a computeValue that returns an array.
A good example is the FunctionArrayAux, with the computeValue below:
```
RealEigenVector
FunctionArrayAux::computeValue()
{
  RealEigenVector v(_var.count());
  const Point & p = isNodal() ? *_current_node : _q_point[_qp];
  for (unsigned int i = 0; i < _var.count(); ++i)
    v(i) = _funcs[i]->value(_t, p);
  return v;
}
```

**Reply by abarun22** [reply]
Hi Guillaume,
The new `ArrayRankTwoAux` i was working on earlier was set to inherit from `NodalPatchRecovery` (as was the case with `RankTwoAux`). As you said this will need inheritance from array AuxKernel to return array variables through computeValue(). I can see that my computeValue() in `ArrayRankTwoAux` now works OK with returning a `RealEigenVector` (which is a 1D vector equivalent). May i ask that if there is anything we could do here to work with a 2D vector (i intend to do a [48,6] return)? I wish that my computeValue() template looks like below.
```
template <bool is_ad> std::vector<std::vector<Real>> ArrayRankTwoAuxTempl<is_ad>::computeValue()
```
What modifications are needed here to make this to work. For your reference here is my array AuxVariable/AuxKernel block in the input file.
```
[AuxVariables]
  [./stress_arav]
    order = CONSTANT
    family = MONOMIAL
    components=48
  [../]
[]

[AuxKernels]
  [./stress_arak]
    type = ArrayRankTwoAux
    rank_two_tensor = stress
    variable = stress_arav
  [../]
[]
```

**Reply by GiudGiud** [reply]
not really. a [48,6] return would mean you are filling some sort of 2D array variable. We dont have that currently in MOOSE. 
ArrayVariable just means 1D array of arbitrary size. Vector variable means 1D vector of size LIBMESH_dimension. 

NodalPatchRecovery inherits from AuxKernel not arrayAuxKernel. 

I think you will need to create an array version of this class to make it work

**Reply by abarun22** [reply]
Hi Guillaume, 
I tried creating an `ArrayAuxKernel` that can handle 2D arrays, but the modifications go very deep in to low level MooseVariable classes and furthermore they are scattered everywhere, making it hard to tune the code for such cases. I start with the following set of definitions in `AuxKernel.h` and tried to connect this with various related entities down the line.
```
typedef std::vector<std::vector<Real>> TwoDRealVector;
typedef AuxKernelTempl<TwoDRealVector> Array2DAuxKernel;
```
This is an extremely complicated task and will need several layers of modifications before we get this to work. A simple workaround will be to use the existing `ArrayKernel` that is type defined for `AuxKernelTempl<RealEigenVector>` which operates on 1D array's. With this approach i can condense my 2D array in to a 1D array say `a[m][n]=a[m*n]` which has already been tested and found working. Wanted to know whether you are OK with this idea or still want to dig deep in to the functions to find compatibility for a 2D array?
Kind regards,
Arun

**Reply by abarun22** [reply]
Also the `components` parameter in `AuxVariables` block will not allow creation of 2D arrays in AuxKernels as it may induce a seg fault. May be this will need some tuning as well for e.g an additional parameter to fix the column size for the array.

**Reply by GiudGiud** [reply]
I would really avoid to modify any core moose code. Try to map to the 1D array at the computeValue level (the one inherited from an ArrayAuxKernel) and no earlier. 
We dont support and currently do not have interest (that I know of, but I m sure someone does) in tensor/matrix variables.

**Reply by abarun22** [reply]
Yes, i think AuxKernels based on 1D array seems to be the way moving forward.
